lib/Dialect/SV/Transforms/RTLLegalizeNames.cpp

```diff
@@ -0,0 +1,198 @@
+//===- RTLCleanup.cpp - RTL Cleanup Pass ----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass renames modules and variables to avoid conflicts
+// with keywords and other declarations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Translation/ExportVerilog.h"
+
+using namespace circt;
+using namespace sv;
+using namespace rtl;
+
+//===----------------------------------------------------------------------===//
+// RTLLegalizeNamesPass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct RTLLegalizeNamesPass
+    : public sv::RTLLegalizeNamesBase<RTLLegalizeNamesPass> {
+  void runOnOperation() override;
+
+private:
+  bool anythingChanged;
+
+  /// Given a module name \p moduleName, return the updated name if it has been
+  /// previously renamed in the table moduleNameTable, else call
+  /// sanitizeName, to get a new name in case of conflict with
+  /// keywords.
+  StringRef getModuleName(StringAttr moduleName) {
+    auto entryIter = moduleNameTable.find(moduleName);
+    if (entryIter != moduleNameTable.end())
+      return entryIter->getSecond();
+    auto updatedName = sanitizeName(moduleName.getValue(), usedModuleNames,
+                                    nextGeneratedNameID);
+    moduleNameTable[moduleName] = updatedName;
+    return updatedName;
+  }
+
+  /// moduleNameTable keeps track of mappings from Module names to updated names
+  /// to resolve keyword conflicts.
+  llvm::DenseMap<Attribute, StringRef> moduleNameTable;
+
+  /// Set of used MLIR module names, to ensure unique names when renaming
+  /// keywords in module names.
+  llvm::StringSet<> usedModuleNames;
+
+  size_t nextGeneratedNameID = 0;
+};
+} // end anonymous namespace
+
+void RTLLegalizeNamesPass::runOnOperation() {
+  anythingChanged = false;
+  auto root = getOperation();
+
+  // We run in two phases. Phase one legalizes module and interface names in the
+  // module globally. Phase two then dives into these structures and performs
+  // local legalization on names that cannot leak out their enclosing module.
+
+  // Register the names of external modules which we cannot rename.
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto extMod = dyn_cast<RTLModuleExternOp>(op)) {
```

Comment:
ESI uses SV interfaces to provide a well-known interface to the outside world. So we rename them, a system with external modules which uses an SV interface will break. I'm pretty sure this is currently a problem, so you're not making it worse just something you might want to fix here. Or I can fix it in a subsequent PR.

Reply:
Same will apply to typedefs in the future.

Reply:
The interface used *within* the extern module will still be renamed (through the symbol table), it's just that the extern module itself won't be renamed. The `walk()` does traverse into extern module's port list though.